source/manuals/programming-languages/python/python.html.md.erb

```diff
@@ -1,106 +1,150 @@
-features such as metaclasses) in these cases we defer to the [Google Python style guide][GPSG].
-We suggest using [PEP8][] and the [Google Python style guide][GPSG] (in order of preference) as references unless
+We follow [PEP 8][]; where [PEP 8][] doesn't express a view (e.g. on the usage of language
+features such as metaclasses) we defer to the [Google Python style guide][GPSG].
+Use these as references unless
 something is explicitly mentioned in the [GDS Python Style Guide][GDSPSG].
 
+The [GDS Python Style Guide][GDSPSG] is the place for detailing GDS specific rules/exemptions to [PEP&nbsp;8][PEP 8]
+or the [Google Python style guide][GPSG]. An example of which is allowing a line length of 120 instead
+of 79.
 
-### Additional rules
-
-The [GDS Python Style Guide][GDSPSG] is the place for detailing GDS specific rules/ exemptions to [PEP8][] or the
-[Google Python style guide][GPSG]. An example of which is allowing a line length of 120 instead of 79.
-
-As always the rules in the [GDS Python Style Guide][GDSPSG] should be followed only in conjunction with the advice
-on consistency on the main [programming languages manual page][GDSPSG].
+As always the rules in the [GDS Python Style Guide][GDSPSG] should be followed only in conjunction
+with the advice on consistency on the main [programming languages manual page][GDSPSG].
 
 If you want to add a new rule or exception please create a pull request against this repo.
 
+## Linting
 
-### Linting
-
-_For more on linting (including standard settings and config) see the [linting][linting] section of this manual._
+_For more on linting (including standard settings and config) see the [linting][] section of this manual._
 
 [Flake8][] is the preferred linting tool. It incorporates:
 
-* [PEP-0008][PEP8] inspired style checks using PyCodeStyle
+* [PEP 8][] inspired style checks using [pycodestyle][]
 * [Complexity][WikiCyclomatic_complexity] checking using the [McCabe project][McCabe]
-* Lint checks using [PyFlakes][]
+* Lint checks using [Pyflakes][]
 
 
-### Dependencies
+## Dependencies
 
-<blockquote>The tools mentioned in this topic are changing quickly, so this
-section would benefit from review in July 2018</blockquote>
+A Python application project typically brings together Python packages from PyPI, with
+others written in-house (or otherwise not distributed via PyPI).
 
-It is important to have a good strategy for specifying your project's Python
-libraries. There are two desirable characteristics:
+There are two ways of specifying dependencies in Python world: as a specific version or
+as an allowable range.
 
-1. Reproducibility
+Different considerations apply to dependency management depending whether you are packaging a library, or creating
+a deployable artifact such as an application or a bundle of scripts: in general, you should only specify specific
+versions at the last responsible moment.
```

Comment:
What is meant by `the last responsible moment`?

Reply:
We mean that you should specify broad version ranges where possible (e.g. so that if you're building a library, people aren't stuck using a specific version just because it happens to claim compatibility with the _exact_ version of some other library you are already using).

But conversely when you pulling together libraries to build an application, then having a precise "known good" set of dependencies that should actually be deployed is very valuable for the reasons given elsewhere in this document.

LRM is a general principle that's googleable but I can see we haven't made its application here particularly obvious, again happy to accept alternative formulations (cc @lfdebrux).

Reply:
can we define it on first use, or link to an external explanation?

Reply:
Using the phrase in this context was a brain-fart of mine; I don't think anyone else has used that form of words when referring to Python dependency management before.

Reply:
Maybe split this into two? The first part explains the *specificity* of dependencies for libraries vs repeatable installation of artifacts. Then a section dicussing *when* you actually decide on your dependency versions. (A definition would be good - I think linking out would be disruptive.)

Reply:
I've added some further text and a diagram, please review if you have a moment